jwst/extract_1d/extract.py

```diff
@@ -1137,6 +1153,134 @@ def shift_by_source_location(location, nominal_location, extract_params):
         if extract_params[params] is not None:
             extract_params[params] += offset
 
+    # Shift the full trace
+    if update_trace and extract_params['trace'] is not None:
+        extract_params['trace'] += offset
+
+
+def _nirspec_trace_from_wcs(shape, bounding_box, wcs_ref, source_xpos, source_ypos):
+    """Calculate NIRSpec source trace from WCS.
+
+    The source trace is calculated by projecting the recorded source
+    positions source_xpos/ypos from the NIRSpec "slit_frame" onto
+    detector pixels.
+
+    Parameters
+    ----------
+    shape : tuple of int
+        2D shape for the full input data array, (ny, nx).
+    bounding_box : tuple
+        A pair of tuples, each consisting of two numbers.
+        Represents the range of useful pixel values in both dimensions,
+        ((xmin, xmax), (ymin, ymax)).
+    wcs_ref : `~gwcs.WCS`
+        WCS for the input data model, containing slit and detector
+        transforms.
+    source_xpos : float
+        Slit position, in the x direction, for the target.
+    source_ypos : float
+        Slit position, in the y direction, for the target.
+
+    Returns
+    -------
+    trace : ndarray of float
+        Fractional pixel positions in the y (cross-dispersion direction)
+        of the trace for each x (dispersion direction) pixel.
+    """
+    x, y = grid_from_bounding_box(bounding_box)
+    nx = int(bounding_box[0][1] - bounding_box[0][0])
+
+    # Calculate the wavelengths in the slit frame because they are in
+    # meters for cal files and um for s2d files
+    d2s = wcs_ref.get_transform("detector", "slit_frame")
+    _, _, slit_wavelength = d2s(x,y)
+
+    # Make an initial array of wavelengths that will cover the wavelength range of the data
+    wave_vals = np.linspace(np.nanmin(slit_wavelength), np.nanmax(slit_wavelength), nx)
+    # Get arrays of the source position in the slit
+    pos_x = np.full(nx, source_xpos)
+    pos_y = np.full(nx, source_ypos)
+
+    # Grab the wcs transform between the slit frame where we know the
+    # source position and the detector frame
+    s2d = wcs_ref.get_transform("slit_frame", "detector")
+
+    # Calculate the expected center of the source trace
+    trace_x, trace_y = s2d(pos_x, pos_y, wave_vals)
+
+    # Interpolate the trace to a regular pixel grid in the dispersion
+    # direction
+    interp_trace = interp1d(trace_x, trace_y, fill_value='extrapolate')
+
+    # Get the trace position for each dispersion element
+    trace = interp_trace(np.arange(nx))
+
+    # Place the trace in the full array
+    full_trace = np.full(shape[1], np.nan)
+    x0 = int(np.ceil(bounding_box[0][0]))
+    full_trace[x0:x0 + nx] = trace
+
+    return full_trace
+
+
+def _miri_trace_from_wcs(shape, bounding_box, wcs_ref, source_ra, source_dec):
```

Comment:
@jemorrison here (_miri_trace_from_wcs) is the relevant function that calculates the trace for MIRI data.  The equivalent function for NIRSpec is similar other than the differences in MIRI/NIRSpec WCS and can be found above (_nirspec_trace_from_wcs).

Reply:
And here is where _miri_trace_from_wcs is called in location_from_wcs, in case it is useful to work backward and see what inputs are provided:  https://github.com/spacetelescope/jwst/blob/60eedabf75c9c35b4d6382ed3784687b670276fd/jwst/extract_1d/extract.py#L1101